src/vision/shiny_future/barbara_enjoys_an_async_sandwich.md

```diff
@@ -0,0 +1,160 @@
+# ✨ Shiny future stories: Barbara enjoys her async-sync-async sandwich :sandwich: 
+
+:::warning
+Alternative titles:
+- Barbara enjoys her async-sync-async sandwich :sandwich: 
+- Barbara recursively blocks
+- Barbara blocks and blocks and blocks
+:::
+
+
+## 🚧 Warning: Draft status 🚧
+
+This is a draft "shiny future" story submitted as part of the brainstorming period. It is derived from what actual Rust users wish async Rust should be, and is meant to deal with some of the challenges that Async Rust programmers face today.
+
+If you would like to expand on this story, or adjust the answers to the FAQ, feel free to open a PR making edits (but keep in mind that, as peoples needs and desires for async Rust may differ greatly, shiny future stories [cannot be wrong]. At worst they are only useful for a small set of people or their problems might be better solved with alternative solutions). Alternatively, you may wish to [add your own shiny vision story][htvsq]!
+
+## The story
+
+Barbara wants to customize a permissions lookup when accepting requests. The library defines a `trait PermitRequest`, to allow the user to define their own rules. Nice!
+
+```rust
+trait PermitRequest {}
+```
+
+She starts small, to get her feet wet.
+
+```rust
+struct Always;
+
+impl PermitRequest for Always {
+    fn permit(&self, _: &Request) -> bool {
+        true
+    }
+}
+```
+
+All requests are permitted! Simple, but now to actually to implement the permissions logic.
+
+One of the basic rules Barbara has is to check the request for the existence of a header, but the function is written as `async`, since Barbara figured it might need to be eventually.
+
+```rust
+async fn req_has_header(req: &Request) -> bool {
+    req.headers().contains_key("open-sesame")
+}
+```
+
+When Barbara goes to implement the `PermitRequest` trait, she realizes a problem: the trait did not think permissions would require an async lookup, so its method is not `async`. Barbara tries the easiest thing first, hoping that she can just block on the future.
+
+```rust
+struct HasHeader;
+
+impl PermitRequest for HasHeader {
+    fn permit(&self, req: &Request) -> bool {
+        task::block_on(req_has_header(req))
```

Comment:
While a `block_on` in std would already go a long way here, given how often this is needed, I wonder if we could do even better and provide a method on futures, so people can just use it like `.await`:

```rust
        req_has_header(req).block()
```

In fact, if it was up to me, I'd even want it as part of the language:

```rust
    req_has_header(req).block
```

Reply:
The original `Future` trait had a `wait` method, which did exactly that. We chose to remove it, because it was _too easy_ for people to reach for instead of just letting the future be chained with others (now I'd say just `.await`). While I want to make it more forgiving in instances when a user _has_ to do this, I'd still want to encourage users to try not to, since it _is_ worse than using `.await`.

Reply:
hmm.. `wait` can very easily be confused with `await` and I can see most programmers new to async, confusing them, but I see your point.

Reply:
Right, I'm referring to `Future` v0.1, which existed before there was async/await syntax. People would often do things like this because of how easy it was:

```rust
first_fut.and_then(|val| {
    // a few lines of whatever
    let foo = do_another_future(val).wait();
    // and other lines
})
```